### PR TITLE
Add cloud environment to MicrosoftTeamsIdentifier

### PR DESCRIPTION
**Pull request by sacheu**
* Add CommunicationCloudEnvironment class
* CommunicationCloudEnvironmentModel is a generated enum
* Add serialize/deserialize code to convert CommunicationCloudEnvironment from/to CommunicationCloudEnvironmentModel
* Add CommunicationCloudEnvironment to MicrosoftTeamsIdentifier

Corresponding .net implementation is  https://github.com/Azure/azure-sdk-for-net/pull/18120

### CODE REVIEW

File: sdk/communication/AzureCommunication/Source/Identifiers.swift
Comment (RezaJooyandeh) [inline sdk/communication/AzureCommunication/Source/Identifiers.swift:110]: `cloudEnvironemt` -> `cloudEnvironment`

Comment (RezaJooyandeh) [reply sdk/communication/AzureCommunication/Source/Identifiers.swift:110]: I see the same spelling issue in other places too. Maybe double check around

File: sdk/communication/AzureCommunication/Source/CommunicationCloudEnvironment.swift
Comment (JoshuaLai) [inline sdk/communication/AzureCommunication/Source/CommunicationCloudEnvironment.swift:32]: Double import can be removed

File: sdk/communication/AzureCommunication/Source/CommunicationCloudEnvironment.swift
Comment (JoshuaLai) [inline sdk/communication/AzureCommunication/Source/CommunicationCloudEnvironment.swift:51]: Great use of equatable

File: sdk/communication/AzureCommunication/Source/CommunicationCloudEnvironment.swift
Comment (JoshuaLai) [inline sdk/communication/AzureCommunication/Source/CommunicationCloudEnvironment.swift:34]: I assume this is @objcMembers cause we expect calling to use this?

File: sdk/communication/AzureCommunication/Source/CommunicationCloudEnvironment.swift
Comment (JoshuaLai) [inline sdk/communication/AzureCommunication/Source/CommunicationCloudEnvironment.swift:60]: These static lets are double defined. I think we can remove them from the class or delete this extension

File: sdk/communication/AzureCommunication/Source/CommunicationCloudEnvironment.swift
Comment (sacheu) [reply sdk/communication/AzureCommunication/Source/CommunicationCloudEnvironment.swift:31]: `CommunicationCloudEnvironment ` is now a property of MicrosoftTeamsUserIdentifier. And MicrosoftTeamsUserIdentifier, is used by calling. 
so that is why I add @objcMembers here too ,

image

File: sdk/communication/AzureCommunication/Source/CommunicationCloudEnvironment.swift
Comment (sacheu) [reply sdk/communication/AzureCommunication/Source/CommunicationCloudEnvironment.swift:29]: Thanks. Removed.

File: sdk/communication/AzureCommunication/Source/Identifiers.swift
Comment (sacheu) [reply sdk/communication/AzureCommunication/Source/Identifiers.swift]: Fixed.

File: sdk/communication/AzureCommunication/Source/CommunicationCloudEnvironment.swift
Comment (sacheu) [reply sdk/communication/AzureCommunication/Source/CommunicationCloudEnvironment.swift:60]: They are not double defined. One set are CommunicationCloudEnvironment, the other ones are **CommunicationIdentifierModel**.  I put them in the same file because **CommunicationIdentifierModel** is generated code. I don't want to put hand written code in the generated file.

File: sdk/communication/AzureCommunication/Source/CommunicationIdentifierSerializer.swift
Comment (DominikMe) [inline sdk/communication/AzureCommunication/Source/CommunicationIdentifierSerializer.swift:52]:
```suggestion
                throw AzureError.client("Can't serialize CommunicationIdentifierModel: isAnonymous is undefined.")
```

File: sdk/communication/AzureCommunication/Source/CommunicationIdentifierSerializer.swift
Comment (DominikMe) [inline sdk/communication/AzureCommunication/Source/CommunicationIdentifierSerializer.swift:100]: We shouldn't throw here but instead 
```suggestion
        return CommunicationCloudEnvironment(cloud.requestString);
```

**Review by DominikMe — COMMENTED** [review]
Looks great Sam, one nit and one change request 📝

File: sdk/communication/AzureCommunication/Source/CommunicationIdentifierSerializer.swift
Comment (sacheu) [reply sdk/communication/AzureCommunication/Source/CommunicationIdentifierSerializer.swift:52]: Good catch.

File: sdk/communication/AzureCommunication/Source/CommunicationIdentifierSerializer.swift
Comment (sacheu) [reply sdk/communication/AzureCommunication/Source/CommunicationIdentifierSerializer.swift:100]: Good catch.

File: sdk/communication/AzureCommunication/Source/CommunicationCloudEnvironment.swift
Comment (JoshuaLai) [reply sdk/communication/AzureCommunication/Source/CommunicationCloudEnvironment.swift:31]: that's too bad I think this would be great to use an enum instead of class

File: sdk/communication/AzureCommunication/Source/Identifiers.swift
Comment (JoshuaLai) [inline sdk/communication/AzureCommunication/Source/Identifiers.swift:107]: I think we should avoid the use of names like `id` since `id` is a reserved type in Objective-C

Comment (JoshuaLai) [reply sdk/communication/AzureCommunication/Source/Identifiers.swift:107]: In most cases we use `public let identifier: String`

File: sdk/communication/AzureCommunication/Source/CommunicationIdentifierSerializer.swift
Comment (JoshuaLai) [inline sdk/communication/AzureCommunication/Source/CommunicationIdentifierSerializer.swift:66]: Great change!

**Review by JoshuaLai — CHANGES_REQUESTED** [review]
Looks good just had some small feedback

**Review by JoshuaLai — APPROVED** [review]

**Review by tjprescott — CHANGES_REQUESTED** [review]
Should this be targeting master or the AzureCommunication feature branch? It is currently targeting master.

**Comment by sacheu** [comment]
@tjprescott  we want this change to be target to Master branch.

**Comment by JoshuaLai** [comment]
> Should this be targeting master or the AzureCommunication feature branch? It is currently targeting master.

@tjprescott Yep this is one we want to merge into Master. However we are going to hold off on merging it in until after the next release.

**Review by tjprescott — APPROVED** [review]

**Review by tjprescott — CHANGES_REQUESTED** [review]
@JoshuaLai marking "Request Changes" until you ping me to merge it.

**Review by tjprescott — APPROVED** [review]

**Pull request merged**
